### ISSUE TITLE
[bitnami/rabbitmq] Prometheus uses Pod name and number for `clustername` in prometheus

### ISSUE THREAD

**Issue by robotaref** [bitnami/charts#15514]
### Name and Version

11.9.1

### What architecture are you using?

None

### What steps will reproduce the bug?

Create a cluster with 3 nodes with simple helm install, and enable prometheus monitoring.
Watch `rabbitmq_identity_info` in prometheus panel.

### Are you using any custom parameters or values?

```  extraConfiguration: |-
    prometheus.return_per_object_metrics = true
```

### What is the expected behavior?

I expect to see the `clustername` using `service_name`, something like:
```
rabbitmq_cluster="rabbit@SERVICE_NAME-headless.NAMESPACE_NAME.svc.CLUSTER_DOMAIN
```

### What do you see instead?

The `clustername` uses the pods name. Something like this:
```
rabbitmq_cluster="rabbit@POD_NAME.SERVICE_NAME-headless.NAMESPACE_NAME.svc.CLUSTER_DOMAIN"
```

**Comment by javsalgar** [comment]
Hi,

Could you specify the exact commands to reproduce the issue in a fresh K8s cluster?

**Comment by marcosbc** [comment]
Hi @robotaref, have you ensured that RabbitMQ configurations within your `values.rabbit.yaml` file are located within the `rabbitmq:` object? That is usually required when configuring as a sub-chart. E.g.:

```
rabbitmq:
  ...
  extraConfiguration: |-
    ...
    prometheus.return_per_object_metrics = true
```

In addition, have you been able to reproduce this issue with a standalone chart?

**Comment by robotaref** [comment]
@marcosbc thanks for your reply.
Yes. that's the way i provided the `config`, just did not want my message to be too long.
And yes again, had the same problem with a standalone version.

**Comment by marcosbc** [comment]
If you set those options to the `extraConfiguration` object, you should find the configuration inside `rabbitmq.conf`. Could you check that this option is present in `/opt/bitnami/rabbitmq/etc/rabbitmq/rabbitmq.conf`?

Just note that, due to how this was implemented in the chart, you cannot set both `configuration` and `extraConfiguration` at the same time. Or, if you do, you will need to ensure that the first one includes the latter via Helm templating (you can see how this is done in the default `values.yaml` file).

**Comment by robotaref** [comment]
@marcosbc
How should this effect node name?
Even without it, the the problem remain.

**Comment by marcosbc** [comment]
Hi @robotaref, I understand now. So you want to use a different node name.

Currently, the node name is hardcoded to a few different values, and I assume it is because this value needs to be resolvable within the internal network. You can find the related logic [here](https://github.com/bitnami/charts/blob/main/bitnami/rabbitmq/templates/statefulset.yaml#L188).

In your case, it seems like you would like to set the `clustering.addressType` chart option to `hostname`, so that it resolves to the address within your Kubernetes cluster.

**Comment by robotaref** [comment]
Hello again @marcosbc 
Maybe i was not clear in my initial question. Address type is set to `hostname` (actually it's left empty and default values is hostname).
The problem is that `clustername` in prometheus metrics like `rabbitmq_identity_info` is dependent to node name.

**Comment by marcosbc** [comment]
Hi @robotaref, one last question to properly understand what might be going on. If you enter the pod with a shell, and check the value in `/opt/bitnami/rabbitmq/etc/rabbitmq/rabbitmq-env.conf`, what value do you see for `NODE_NAME`? Is it the expected one or wrong one?

If it were the wrong one, it's definitely related with how you are deploying your chart. If it is the expected one, then there could be some issues with RabbitMQ configuration that is causing the cluster name not to be properly set.

**Comment by robotaref** [comment]
Hi @marcosbc,
Node name is something like this, which is what's expected:
```
HOME=/opt/bitnami/rabbitmq/.rabbitmq
NODE_PORT=5672
NODENAME=rabbit@POD_NAME.SERVICE_NAME-rabbitmq-headless.NAME_SPACE.svc.CLUSTER_DOMAIN
```
but there is no trace of cluster name in here as another value.

**Comment by marcosbc** [comment]
Have you also checked if the cluster name appears in the shell environment? E.g. executing `env` in the shell session.

Please also do a recursive `grep` in the Redis directories (`/bitnami/redis` and `/opt/bitnami/redis`) in case the configuration appears in any place there. E.g. `grep -r CLUSTER_NAME /bitnami/redis /opt/bitnami/redis` (and replacing `CLUSTER_NAME` with a pattern for what you are seeing)

If it does not appear at any location, then it can be that the container is missing a configuration.

**Comment by leaqui** [comment]
Although it would be great to solve this problem, I add two workarounds:
- [https://github.com/bitnami/charts/issues/12161](https://github.com/bitnami/charts/issues/12161)
- https://groups.google.com/g/rabbitmq-users/c/A8v8XSmnCnA?pli=1

Both change cluster name after start, the first one using `rabbitmq-init-scripts` an the second, manually.

**Comment by leaqui** [comment]
But it's not solved.

`clustername` is wrong until you change it (manually or by script), each pod have different value.

Someone know where is specified using pod name at cluster name?

**Comment by corico44** [comment]
@leaqui You can find this logic [here.](https://github.com/bitnami/charts/blob/main/bitnami/rabbitmq/templates/statefulset.yaml#L188)

**Comment by leaqui** [comment]
Thanks @corico44 for your reply.

But this is RabbitMQ node name definition, not cluster name.

**Comment by corico44** [comment]
> But it's not solved.
> 
> `clustername` is wrong until you change it (manually or by script), each pod have different value.
> 
> Does anyone know where is specified to use pod name at cluster name?

@leaqui How did you modify it manually or by script? I can't find the parameter you mean.

**Comment by leaqui** [comment]
@corico44 you can find the init script way here: https://github.com/bitnami/charts/issues/12161#issue-1351204314

and manually you can use the `rabbitmqctl set_cluster_name` command to set a cluster name after forming your cluster.

**Comment by leaqui** [comment]
just to not be closed

**Comment by philipp-durrer-jarowa** [comment]
I've just came across this issue to, will try to use the init script linked above and report back.

**Comment by leaqui** [comment]
Please, don't close

**Comment by corico44** [comment]
Hello! @philipp-durrer-jarowa any updates on your workaround?
Also, I have opened an internal ticket to investigate the rabbitmq configuration and see if the clustername is configured. Any news will be reported in this issue.

**Comment by philipp-durrer-jarowa** [comment]
I'm currently struggling to get the nodes back online when they've all been stopped. The first pod is crashlooping:
```
$ kc logs jarowa-events-rabbitmq-0
rabbitmq 08:45:41.30 
rabbitmq 08:45:41.31 Welcome to the Bitnami rabbitmq container
rabbitmq 08:45:41.31 Subscribe to project updates by watching https://github.com/bitnami/containers
rabbitmq 08:45:41.31 Submit issues and feature requests at https://github.com/bitnami/containers/issues
rabbitmq 08:45:41.31
rabbitmq 08:45:41.31 INFO  ==> ** Starting RabbitMQ setup **
rabbitmq 08:45:41.33 INFO  ==> Validating settings in RABBITMQ_* env vars..
rabbitmq 08:45:41.35 INFO  ==> Initializing RabbitMQ...
rabbitmq 08:45:41.39 INFO  ==> Persisted data detected. Restoring...
rabbitmq 08:45:41.39 INFO  ==> Loading user's custom files from /docker-entrypoint-initdb.d ...
Error: unable to perform an operation on node 'rabbit@jarowa-events-rabbitmq-0.jarowa-events-rabbitmq-headless.jarowa-enable.svc.cluster.local'. Please see diagnostics information and suggestions below.

Most common reasons for this are:

 * Target node is unreachable (e.g. due to hostname resolution, TCP connection or firewall issues)
 * CLI tool fails to authenticate with the server (e.g. due to CLI tool's Erlang cookie not matching that of the server)
 * Target node is not running

In addition to the diagnostics info below:

 * See the CLI, clustering and networking guides on https://rabbitmq.com/documentation.html to learn more
 * Consult server logs on node rabbit@jarowa-events-rabbitmq-0.jarowa-events-rabbitmq-headless.jarowa-enable.svc.cluster.local
 * If target node is configured to use long node names, don't forget to use --longnames with CLI tools

DIAGNOSTICS
===========

attempted to contact: ['rabbit@jarowa-events-rabbitmq-0.jarowa-events-rabbitmq-headless.jarowa-enable.svc.cluster.local']

rabbit@jarowa-events-rabbitmq-0.jarowa-events-rabbitmq-headless.jarowa-enable.svc.cluster.local:
  * connected to epmd (port 4369) on jarowa-events-rabbitmq-0.jarowa-events-rabbitmq-headless.jarowa-enable.svc.cluster.local
  * epmd reports: node 'rabbit' not running at all
                  no other nodes on jarowa-events-rabbitmq-0.jarowa-events-rabbitmq-headless.jarowa-enable.svc.cluster.local
  * suggestion: start the node

Current node details:
 * node name: 'rabbitmqcli-734-rabbit@jarowa-events-rabbitmq-0.jarowa-events-rabbitmq-headless.jarowa-enable.svc.cluster.local'
 * effective user's home directory: /opt/bitnami/rabbitmq/.rabbitmq
 * Erlang cookie hash: Dc3jWh3wPX8AazVHQeyiGg==
```

**Comment by philipp-durrer-jarowa** [comment]
Ah ok, I've found the problem, on a subsequent start of the pod the init scripts get run before the cluster is available, which makes it exit with a non-zero exit code and therefore crashs the pod. I guess I can just modify the script to make sure that the cluster is running as it get's executed correctly on first startup.

**Comment by philipp-durrer-jarowa** [comment]
Comparing to the first start, where rabbitMQ get's started in the background before the init scripts are run:
```
rabbitmq 09:12:26.93 INFO  ==> ** Starting RabbitMQ setup **
rabbitmq 09:12:26.94 INFO  ==> Validating settings in RABBITMQ_* env vars..
rabbitmq 09:12:26.97 INFO  ==> Initializing RabbitMQ...
rabbitmq 09:12:27.01 INFO  ==> Starting RabbitMQ in background...
rabbitmq 09:12:43.16 INFO  ==> Loading user's custom files from /docker-entrypoint-initdb.d ...
Status of node rabbit@jarowa-events-rabbitmq-0.jarowa-events-rabbitmq-headless.jarowa-enable.svc.cluster.local ...
...
Interface: [::], port: 15672, protocol: http, purpose: HTTP API
Interface: [::], port: 9419, protocol: http/prometheus, purpose: Prometheus exporter API over HTTP
Interface: [::], port: 25672, protocol: clustering, purpose: inter-node and CLI tool communication
Interface: [::], port: 5672, protocol: amqp, purpose: AMQP 0-9-1 and AMQP 1.0
Setting cluster name to rabbit@narnia-pool-events ...
rabbitmq 09:12:44.80 INFO  ==> Stopping RabbitMQ...
rabbitmq 09:12:52.68 INFO  ==> ** RabbitMQ setup finished! **

rabbitmq 09:12:52.69 INFO  ==> ** Starting RabbitMQ **
...
```

So I've change the script to this now and it works:
```
initScripts:
    my_init_script.sh: |
      #!/bin/sh
      rabbitmqctl status && rabbitmqctl set_cluster_name rabbit@{{ .Values.cluster.name }}
      exit 0
```

And the Grafana graphs are correctly aggregated:
![image](https://github.com/bitnami/charts/assets/135346089/22c439d6-7187-4174-bf20-715b632e1ca8)

**Comment by leaqui** [comment]
Hi @carrodher, this is not solved.

Did you find where is specified to use pod name at cluster name?

**Comment by simi** [comment]
I can confirm, this is still problem.

**Comment by philipp-durrer-jarowa** [comment]
@leaqui @simi did you see my comment? https://github.com/bitnami/charts/issues/15514#issuecomment-1670966093 you have to add the script as outlined by me to the initScripts sections in your values.yaml file.

**Comment by simi** [comment]
> @leaqui @simi did you see my comment? [#15514 (comment)](https://github.com/bitnami/charts/issues/15514#issuecomment-1670966093) you have to add the script as outlined by me to the initScripts sections in your values.yaml file.

Sorry for late reply. Thanks for the suggestion. It works, but clearly only for new clusters. When added to already created cluster, this has no effect. I was able to change cluster name manually for running clusters and add this in case of re-creation. Would it make sense to make this built-in chart itself?

**Comment by leaqui** [comment]
> @leaqui @simi did you see my comment? [#15514 (comment)](https://github.com/bitnami/charts/issues/15514#issuecomment-1670966093) you have to add the script as outlined by me to the initScripts sections in your values.yaml file.

Yes @philipp-durrer-jarowa, I've seen it, thanks for replying!

It's a good workaround but it doesn't solve the original problem.

**Comment by juan131** [comment]
Hi everyone

It seems the `rabbitmq_cluster_permanent_id` is properly configured (stable & unique identifier). These are the steps that I followed:

- Install the chart enabling metrics:

```bash
helm install rabbitmq oci://registry-1.docker.io/bitnamicharts/rabbitmq --set replicaCount=3 --set metrics.enabled=true
```

- Wait for replicas to be ready and inspect the metrics:

```console
$ kubectl exec -it rabbitmq-0 -- curl -sL http://127.0.0.1:9419/metrics | grep rabbitmq_identity_info
# TYPE rabbitmq_identity_info untyped
# HELP rabbitmq_identity_info RabbitMQ node & cluster identity info
rabbitmq_identity_info{rabbitmq_node="rabbit@rabbitmq-0.rabbitmq-headless.default.svc.cluster.local",rabbitmq_cluster="rabbit@rabbitmq-0.rabbitmq-headless.default.svc.cluster.local",rabbitmq_cluster_permanent_id="rabbitmq-cluster-id-bCGqz5j0-1SdZxf7J4eYFw"} 1
$ kubectl exec -it rabbitmq-1 -- curl -sL http://127.0.0.1:9419/metrics | grep rabbitmq_identity_info
# TYPE rabbitmq_identity_info untyped
# HELP rabbitmq_identity_info RabbitMQ node & cluster identity info
rabbitmq_identity_info{rabbitmq_node="rabbit@rabbitmq-1.rabbitmq-headless.default.svc.cluster.local",rabbitmq_cluster="rabbit@rabbitmq-1.rabbitmq-headless.default.svc.cluster.local",rabbitmq_cluster_permanent_id="rabbitmq-cluster-id-bCGqz5j0-1SdZxf7J4eYFw"} 1
$ kubectl exec -it rabbitmq-2 -- curl -sL http://127.0.0.1:9419/metrics | grep rabbitmq_identity_info
# TYPE rabbitmq_identity_info untyped
# HELP rabbitmq_identity_info RabbitMQ node & cluster identity info
rabbitmq_identity_info{rabbitmq_node="rabbit@rabbitmq-2.rabbitmq-headless.default.svc.cluster.local",rabbitmq_cluster="rabbit@rabbitmq-2.rabbitmq-headless.default.svc.cluster.local",rabbitmq_cluster_permanent_id="rabbitmq-cluster-id-bCGqz5j0-1SdZxf7J4eYFw"} 1
```

As you can see, the value reported for `rabbitmq_cluster_permanent_id` is `rabbitmq-cluster-id-bCGqz5j0-1SdZxf7J4eYFw` on every single pod. Therefore, couldn't you use it in your Prometheus/Grafana queries?

**Comment by philipp-durrer-jarowa** [comment]
The most popular Dashboard https://grafana.com/grafana/dashboards/10991-rabbitmq-overview/ is usuing the rabbitmq_cluster label as selector. Also in my opinion the rabbitmq_cluster label should be the same accross all nodes in a cluster.

**Comment by simi** [comment]
@juan131 is there any way to set this to custom value? Should be this recommended to be used at https://github.com/rabbitmq/rabbitmq-server/blob/a2ec45208857bd4743ad12d84f094666f745436d/deps/rabbitmq_prometheus/docker/grafana/dashboards/RabbitMQ-Overview.json? I'm not sure it could be used, since this label is not present at interesting metrics like `rabbitmq_connections`.

**Comment by juan131** [comment]
Hi everyone! I just created a PR addressing this issue, could you please give it a try and let us know if it addresses your problems? Thanks in advance.

**Comment by juan131** [comment]
Hi everyone!!

My changes were merged and a new chart version (**12.6.0**) will be released within the next few hours including them. Please note this new version uses the "common.names.fullname" template to set a stable name for `rabbitmq_cluster` property by default, but it can also be customized to any value using the `clustering.name` property. These changes should make the chart compatible with the official Grafana dashboard.

Please do not hesitate to let us know if you find any issue with this new release.
